### PR TITLE
Refactoring into modules and traits

### LINKED ISSUE #19: Run clippy on workspace members when --no-default-features is passed
The --no-default-features flag cannot be used in clippy with a workspace directory.
cargo-scout should therefore walk through each workspace's members.

In order to avoid increasing run time, we might find a way to _not_ run clippy on workspace members that aren't targeted by the git diff

### PR DESCRIPTION
**Pull request by o0Ignition0o**
Fixes #19 and #13 

Refactored functionalities into separate folders. 
Define behaviors in traits. 
Added folder iteration support within clippy.

### CODE REVIEW

**Comment by o0Ignition0o** [comment]
@ebroto there's quite a lot of changes here, so I'm going to rush for the merge to avoid having too many conflicts with pending PRs.

Could you have a look at the code here if you have a moment? (some tests are missing for now, but I would really like to land it asap so other PRs aren't stalled)

File: src/linter/clippy.rs [hunk 1]
```diff
@@ -159,9 +122,9 @@ pub fn lints(clippy_output: &str) -> Vec<Lint> {
 mod tests {
     #[test]
     fn test_set_verbose() {
-        use crate::clippy::Linter;
+        use crate::linter::clippy::Clippy;
```
Comment (ebroto) [inline src/linter/clippy.rs]: Maybe we could simplify those imports with an `use super::*;` at the mod level?

File: src/scout/mod.rs [hunk 2]
```diff
@@ -1,24 +1,87 @@
-use crate::clippy;
-use crate::git;
+use crate::git::Section;
+use crate::linter::Lint;
+use crate::linter::Span;
+use std::convert::TryInto;
+pub struct Builder {
```
Comment (ebroto) [inline src/scout/mod.rs:5]: missing newline

File: src/scout/mod.rs [hunk 3]
```diff
@@ -1,24 +1,87 @@
...
+        }
+    }
+}
 
 // Check if clippy_lint and git_section have overlapped lines
-fn lines_in_range(clippy_lint: &clippy::Span, git_section: &git::Section) -> bool {
+fn lines_in_range(clippy_lint: &Span, git_section: &Section) -> bool {
     // If git_section.line_start is included in the clippy_lint span
-    clippy_lint.line_start <= git_section.line_start && git_section.line_start <= clippy_lint.line_end ||
+    clippy_lint.line_start <= git_section.line_start.try_into().unwrap() && git_section.line_start <= clippy_lint.line_end.try_into().unwrap() ||
```
Comment (ebroto) [inline src/scout/mod.rs]: Why `try_into()` and `unwrap()` instead of just `into()`?

File: src/scout/mod.rs [hunk 4]
```diff
@@ -1,24 +1,87 @@
...
+            project: None,
+            linter: None,
+        }
+    }
+
+    pub fn set_project_config(
+        &mut self,
+        project: impl crate::project::Config + 'static,
+    ) -> &mut Self {
+        self.project = Some(Box::new(project));
```
Comment (ebroto) [inline src/scout/mod.rs]: Why not taking the `Box` directly as a parameter? I think it would signal better the intent

**Review by ebroto — CHANGES_REQUESTED** [review]
I left some comments. Bear in mind that I'm a newcomer to Rust, so disregard any weird suggestion I may have done :)

File: src/linter/clippy.rs (hunk 1 above)
Comment (o0Ignition0o) [reply src/linter/clippy.rs]: totally !

File: src/scout/mod.rs (hunk 3 above)
Comment (o0Ignition0o) [reply src/scout/mod.rs]: Woops some leftover from when I was bikeshedding ^^

File: src/scout/mod.rs (hunk 4 above)
Comment (o0Ignition0o) [reply src/scout/mod.rs]: It would indeed, I'm a bit sad I have to put it in boxes because an impl is not Sized but yeah why not :)

**Pull request merged**